src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,374 @@
+{
+  const auto it = attributes.find("password");
+  if (it == attributes.end()) {
+    return nonstd::nullopt;
+  }
+  return it->second;
+}
+
+RedisStorage::RedisStorage(const Url& url, const AttributeMap& attributes)
+  : m_url(url),
+    m_connect_timeout(parse_connect_timeout(attributes)),
+    m_operation_timeout(parse_operation_timeout(attributes)),
+    m_username(parse_username(attributes)),
+    m_password(parse_password(attributes))
+{
+  m_prefix = "ccache"; // TODO: attribute
+  m_context = nullptr;
+  m_connected = false;
+  m_invalid = false;
+}
+
+RedisStorage::~RedisStorage()
+{
+  disconnect();
+  if (m_context) {
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+}
+
+int
+RedisStorage::connect()
+{
+  if (m_connected) {
+    return REDIS_OK;
+  }
+  if (m_invalid) {
+    return REDIS_ERR;
+  }
+
+  if (m_context) {
+    if (redisReconnect(m_context) == REDIS_OK) {
+      m_connected = true;
+      return REDIS_OK;
+    }
+    LOG("Redis reconnect err: {}", m_context->errstr);
+    redisFree(m_context);
+    m_context = nullptr;
+  }
+
+  ASSERT(m_url.scheme() == "redis");
+  std::string host = m_url.host();
+  std::string port = m_url.port();
+  std::string sock = m_url.path();
+  if (m_connect_timeout) {
+    LOG("Redis connect timeout {}", timeval_to_string(*m_connect_timeout));
+  }
+  if (!host.empty()) {
+    int p = port.empty() ? 6379 : std::stoi(port);
+    if (m_connect_timeout) {
+      m_context = redisConnectWithTimeout(host.c_str(), p, *m_connect_timeout);
+    } else {
+      m_context = redisConnect(host.c_str(), p);
+    }
+  } else if (!sock.empty()) {
+    if (m_connect_timeout) {
+      m_context = redisConnectUnixWithTimeout(sock.c_str(), *m_connect_timeout);
+    } else {
+      m_context = redisConnectUnix(sock.c_str());
+    }
+  } else {
+    LOG("Redis invalid url: {}", m_url.str());
+    m_invalid = true;
+    return REDIS_ERR;
+  }
+
+  if (!m_context) {
+    LOG("Redis connect {} err NULL", m_url.str());
+    m_invalid = true;
+    return REDIS_ERR;
+  } else if (m_context->err) {
+    LOG("Redis connect {} err: {}", m_url.str(), m_context->errstr);
+    m_invalid = true;
+    return m_context->err;
+  } else {
+    if (m_context->connection_type == REDIS_CONN_TCP) {
+      LOG(
+        "Redis connect tcp {}:{} OK", m_context->tcp.host, m_context->tcp.port);
+    }
+    if (m_context->connection_type == REDIS_CONN_UNIX) {
+      LOG("Redis connect unix {} OK", m_context->unix_sock.path);
+    }
+    m_connected = true;
+
+    if (m_operation_timeout) {
+      LOG("Redis timeout {}", timeval_to_string(*m_operation_timeout));
+      if (redisSetTimeout(m_context, *m_operation_timeout) != REDIS_OK) {
+        LOG_RAW("Failed to set timeout");
+      }
+    }
+
+    // TODO: break out AUTH to separate function
+    if (m_password) {
+      std::string username = m_username ? *m_username : "default";
+      LOG("Redis AUTH {} {}", username, "*****"); // don't log m_password !!!
```

Comment:
Note that the password will be logged as part of the configuration dump anyway. Unless we do something about that, of course, but my [current suggestion is to not do it](https://github.com/ccache/ccache/pull/881#discussion_r664074292).

Reply:
I think it should be masked in case someone decides to attach the log as part of a bug report or something, but your call.

Reply:
I agree that (all other things being equal) it would be better to mask passwords, but I don't want to let `Config` know about what is a password and not. I guess we could enhance the backend API somehow and let backends parse configuration early, communicating back to Config, but my initial reaction to this was to not solve this problem right now since there are many other things up in the air already. But I'll sleep on it.

Reply:
It's a boolean now, so it could go either way in the future depending on what you come up with.

Reply:
I wrote:

> I agree that (all other things being equal) it would be better to mask passwords [...] But I'll sleep on it.

Addressed in 873384667a5e8ba0e688303e18f34196ab3b57c6.